src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java

```diff
@@ -0,0 +1,213 @@
+    return this.tableName;
+  }
+
+  /** @return List of column names specified in COPY statement, if provided. */
+  public List<String> getCopyColumnNames() {
+    return this.copyColumnNames;
+  }
+
+  /** @return Format type specified in COPY statement, if provided. */
+  public String getFormatType() {
+    return this.formatType;
+  }
+
+  /** @return Null string specified in COPY statement, if provided. */
+  public String getNullString() {
+    return this.nullString;
+  }
+
+  /** @return Delimiter character specified in COPY statement, if provided. */
+  public char getDelimiterChar() {
+    return this.delimiterChar;
+  }
+
+  /** @return Escape character specified in COPY statement, if provided. */
+  public char getEscapeChar() {
+    return this.escapeChar;
+  }
+
+  /** @return Quote character specified in COPY statement, if provided. */
+  public char getQuoteChar() {
+    return this.quoteChar;
+  }
+
+  /** @return True if copy data contains a header, false otherwise. */
+  public boolean hasHeader() {
+    return this.hasHeader;
+  }
+
+  public int getFormatCode() {
+    return this.formatCode;
+  }
+
+  public MutationBuilder getMutationBuilder() {
+    return this.mutationBuilder;
+  }
+
+  private void verifyCopyColumns() throws SQLException {
+    if (this.copyColumnNames.size() > this.tableColumns.size()) {
+      throw new SQLException("Number of copy columns provided exceed table column count");
+    }
+    LinkedHashMap<String, TypeCode> tempTableColumns = new LinkedHashMap<>();
+    // Verify that every copy column given is a valid table column name
+    for (String copyColumn : this.copyColumnNames) {
+      if (!this.tableColumns.containsKey(copyColumn)) {
+        throw new SQLException(
+            "Column \"" + copyColumn + "\" of relation \"" + this.tableName + "\" does not exist");
+      }
+      // Update table column ordering with the copy column ordering
+      tempTableColumns.put(copyColumn, tableColumns.get(copyColumn));
+    }
+    this.tableColumns = tempTableColumns;
+  }
+
+  private static TypeCode parseSpannerDataType(String columnType) {
+    // Eliminate size modifiers in column type (e.g. STRING(100), BYTES(50), etc.)
+    int index = columnType.indexOf("(");
+    columnType = (index > 0) ? columnType.substring(0, index) : columnType;
+    TypeCode type = TypeCode.valueOf(columnType);
+    if (type == null) {
+      throw new IllegalArgumentException(
+          "Unrecognized or unsupported column data type: " + columnType);
+    }
+    return type;
+  }
+
+  private void queryInformationSchema() throws SQLException {
+    Map<String, TypeCode> tableColumns = new LinkedHashMap<>();
+    PreparedStatement statement =
+        this.connection.prepareStatement(
+            "SELECT ?, ? FROM information_schema.columns WHERE table_name = '?'");
+    statement.setString(1, COLUMN_NAME);
+    statement.setString(2, SPANNER_TYPE);
+    statement.setString(3, this.getTableName());
```

Comment:
I don't think this works.

1. Column names cannot be specified as query parameters. This query will now return the literal strings 'COLUMN_NAME' and 'SPANNER_TYPE' instead of the actual column names and types.
2. The `'?'` is interpreted as a string literal and not as a query parameter.


```suggestion
            "SELECT column_name, spanner_type FROM information_schema.columns WHERE table_name = ?");
    statement.setString(1, this.getTableName());
```

Reply:
I updated it.